File: ruby/easy/diameter_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
    return find_diameter(root)[:diameter]
end

def find_diameter(root)
    return { diameter: 0, depth: 0 } unless root
    left = find_diameter(root.left)
    right = find_diameter(root.right)
    return { diameter: [left[:diameter], right[:diameter], (left[:depth] + right[:depth])].max, 
                depth: [left[:depth], right[:depth]].max + 1 };
end

File: javascript/easy/diameterOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    return findDiameter(root).diameter;
};

var findDiameter = function(root) {
    if(root === null) return { diameter: 0, depth: 0 };
    const left = findDiameter(root.left);
    const right = findDiameter(root.right);
    const depth = Math.max(left.depth, right.depth) + 1;
    const diameter = Math.max(left.diameter, right.diameter, (left.depth+right.depth));
    return { diameter: diameter, depth: depth };
};
